apply/expand.go

```diff
@@ -0,0 +1,99 @@
+	switch cluster.Spec.Provider {
+	case common.BAREMETAL:
+		if err := PreProcessIPList(scalingArgs); err != nil {
+			return err
+		}
+		if (!IsIPList(scalingArgs.Nodes) && scalingArgs.Nodes != "") || (!IsIPList(scalingArgs.Masters) && scalingArgs.Masters != "") {
+			return fmt.Errorf(" Parameter error: The current mode should submit iplist！")
+		}
+		if scalingArgs.Nodes != "" && IsIPList(scalingArgs.Nodes) && scalingArgs.Masters != "" && IsIPList(scalingArgs.Masters) {
+			margeNodes := append(cluster.Spec.Nodes.IPList, strings.Split(scalingArgs.Nodes, ",")...)
+			margeMasters := append(cluster.Spec.Masters.IPList, strings.Split(scalingArgs.Masters, ",")...)
+			cluster.Spec.Nodes.IPList = removeIPListDuplicatesAndEmpty(margeNodes)
+			cluster.Spec.Masters.IPList = removeIPListDuplicatesAndEmpty(margeMasters)
+			return nil
+		}
+		if scalingArgs.Nodes == "" && scalingArgs.Masters != "" && IsIPList(scalingArgs.Masters) {
+			margeMasters := append(cluster.Spec.Masters.IPList, strings.Split(scalingArgs.Masters, ",")...)
+			cluster.Spec.Masters.IPList = removeIPListDuplicatesAndEmpty(margeMasters)
+			return nil
+		}
+		if scalingArgs.Nodes != "" && scalingArgs.Masters == "" && IsIPList(scalingArgs.Nodes) {
+			margeNodes := append(cluster.Spec.Nodes.IPList, strings.Split(scalingArgs.Nodes, ",")...)
+			cluster.Spec.Nodes.IPList = removeIPListDuplicatesAndEmpty(margeNodes)
+			return nil
+		}
+		return fmt.Errorf(" Parameter error: The current mode should submit iplist！")
+	case common.AliCloud:
+		if (!IsNumber(scalingArgs.Nodes) && scalingArgs.Nodes != "") || (!IsNumber(scalingArgs.Masters) && scalingArgs.Masters != "") {
+			return fmt.Errorf(" Parameter error: The number of join masters or nodes that must be submitted to use cloud service！")
+		}
+		if scalingArgs.Nodes != "" && IsNumber(scalingArgs.Nodes) && scalingArgs.Masters != "" && IsNumber(scalingArgs.Masters) {
+			cluster.Spec.Masters.Count = strconv.Itoa(StrToInt(cluster.Spec.Masters.Count) + StrToInt(scalingArgs.Masters))
+			cluster.Spec.Nodes.Count = strconv.Itoa(StrToInt(cluster.Spec.Nodes.Count) + StrToInt(scalingArgs.Nodes))
+			return nil
+		}
+		if scalingArgs.Nodes == "" && scalingArgs.Masters != "" && IsNumber(scalingArgs.Masters) {
+			cluster.Spec.Masters.Count = strconv.Itoa(StrToInt(cluster.Spec.Masters.Count) + StrToInt(scalingArgs.Masters))
+			return nil
+		}
+		if scalingArgs.Nodes != "" && scalingArgs.Masters == "" && IsNumber(scalingArgs.Nodes) {
+			cluster.Spec.Nodes.Count = strconv.Itoa(StrToInt(cluster.Spec.Nodes.Count) + StrToInt(scalingArgs.Nodes))
+			return nil
+		}
+		return fmt.Errorf(" Parameter error: The number of join masters or nodes that must be submitted to use cloud service！")
+	default:
+		return fmt.Errorf(" clusterfile provider type is not found ！")
+	}
+}
+
+func StrToInt(str string) int {
+	num, err := strconv.Atoi(str)
+	if err != nil {
+		logger.Error("String to digit conversion failed:", err)
```

Comment:
errors occurred but return 0, the caller will set master and nodes count as 0, this is horrible.